### ISSUE TITLE
Cannot push to docker.io; get "404 Not Found" error in the middle of push

### ISSUE THREAD

**Issue by ahgraber** [akhilerm/tag-push-action#88]
Hello!

I'm trying to push images from ghcr.io to docker.io and I get the following error:

```
elapsed: 12.0s                                                                    total:  456.0  (38.0 MiB/s)                                      
ctr: failed commit on ref "manifest-sha256:afbf5e46ea9d4957bbde55c336f42f5a229a28ab8295fadff6ccc1d6478d7a7d": unexpected status: 404 Not Found
panic: exit status 1

goroutine 1 [running]:
main.main()
	/go/src/github.com/tonistiigi/repo-copy/main.go:21 +0x157
Error: The process '/usr/bin/docker' failed with exit code 2
```

Here's the relevant snippet of my github actions:
```
      # copy image from ghcr to dockerhub
      - name: Push image to alternate registries
        # push to registry on main branch only
        if: github.ref == 'refs/heads/main'
        uses: akhilerm/tag-push-action@v1.1.0
        with:
          src: ghcr.io/ninerealmlabs/base_env:python-${{ steps.tag.outputs.tag }}-${{ steps.sha.outputs.sha }}
          dst: |
            docker.io/ninerealmlabs/base_env:python-${{ steps.tag.outputs.tag }}
            docker.io/ninerealmlabs/base_env:python-${{ steps.tag.outputs.tag }}-${{ steps.sha.outputs.sha }}
```

**Comment by akhilerm** [comment]
@ahgraber Can you rerun the action/workflow, to see if its a flaky issue?

**Comment by ahgraber** [comment]
I've tried rerunning the workflow and pushing a new commit to trigger; happens regardless.

Bit more log (if it helps):
```
Run akhilerm/tag-push-action@v1.1.0
/usr/bin/docker run --rm -i -v /home/runner/.docker/config.json:/root/.docker/config.json --network host tonistiigi/repo-copy:latest ghcr.io/ninerealmlabs/base_env:python-3.7-main-86707c6 docker.io/ninerealmlabs/base_env:3.7 docker.io/ninerealmlabs/base_env:python-3.7-86707c6
Unable to find image 'tonistiigi/repo-copy:latest' locally
latest: Pulling from tonistiigi/repo-copy

...

elapsed: 17.4s                                                                    total:  724.2  (41.6 MiB/s)                                      
ctr: failed commit on ref "manifest-sha256:a76d2383cbd449c817ef08552311bf4d5a75a6df74cd91a59644f09edd923329": unexpected status: 404 Not Found
panic: exit status 1

goroutine 1 [running]:
main.main()
	/go/src/github.com/tonistiigi/repo-copy/main.go:21 +0x157
Error: The process '/usr/bin/docker' failed with exit code 2
```

To be fair, these are large images (2+ GB).  Is it possible that there's a timeout issue and github is closing the connection?  
It works fine when building & pushing locally to docker directly, so its not a docker issue or a problem with the repo or tags...

**Comment by ahgraber** [comment]
I take my question back -- it looks like it uploaded the full image _except_ for the manifest, and crashed when uploading the manifest:

```
manifest-sha256:566a6a529dde04246b54d2529813c07a41aab856f90e2d4f54e976388867083d: waiting        |--------------------------------------| 
layer-sha256:c549ccf8d472c3bce9ce02e49c62b8f6cbc736ea2b8ba812a1ae9390c69d0b71:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:e206fcfcaa68aada0f94e858f8d424f91af0e0c546a729bc596bcd42b3abc3c3:    done           |++++++++++++++++++++++++++++++++++++++| 
config-sha256:94c8dc4d4b3105ad81fe1b32d3b6efcb76cdb1c504f210a418360cd607f8aa56:   done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:de26e676faa8dace0626a4b8f057d0ffc7e5139850cda2ab65b8ba0cf807a9b5:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:6c65c50510f39c06f51e7754c41825c087f5b8723aa432478e3f882dc7df5642:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:a89eb75169a1f0fb62c225c407f426d720913c741535cb49eedcb9502e2da544:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:c18636dc46c633cdc8d8950f155a7198ce1ad471661004a9d1b2f3c7cccae942:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:4f71d15c3a446a661c8e9ab34b01fe92bb2883b38da3cec9c950e6c09aa42806:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:7c5f5d979663fd14cba081a67be95348f9cc02e733f225ad804985ba2d46fb28:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:a63f5c102e3f9e02c34d75263760a275a853532c74f1d3167285f8d50f53c5a9:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:ec8b60697d022740e75072ce6c66e9f62f7c6a092adb58d67f7bcbf7073fbb4a:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:4ef8cbbd0ce3a5179cb0bf1b811c6085bfc75c884a72ab785781463ea98b853a:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:4e24d4e20fa90522001cff568936a678858181715a8ec17d95aef35d16c9ac03:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:e888fd7dc15b0eb45fd613496e238da840cdf4cfbdd87d97f450883af236a420:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:014b3810d91b80a9b7edcbce91e7645e25cd5c2c44d443f4fc6ac6552d5a7815:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:fd1d80db8223fbd21a5b7e409e67392944722a79c5c4c15f9c03545ce27fc753:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:853f1ee8487276eb37aafedb51190aa8a7d939f2d881d714a228ab9f8e1989d3:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:35a341dd8cc8619373053d7ee65e31727215c7598a7b8b7639cab5bb60ebcd6c:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:b1ab788b1035e3083f31903897e814d6fb66b3a1a06beccbd41b1d098a3c6dbb:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:0139f532dea9b33551a7c7cf03ff68992acdd50722211374ac2b4524ab3b62f2:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:81b822a9acd6ac46e4a62639e5bbb3492dcf0840bf6dbb0f144a2caf89db0af8:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:38793e44da0b76e16afca557876128b16dd3972c3db143039e585868d04ffc13:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:fa2628cfbdaabed19236bb2f5732eb05897fbe607204c3782ed6789fa804cee1:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:667d2e5ed3c8828002d782c84374d74c91a61eace367510e107f39d3d67c3b72:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:ae12f1b887c3bbba6da00f220e550bec8cca69f61b1e208066da2cfdaff6b8ba:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:4f546916dcf91371fc0f6fe56058dae34b4230c95852c0371bc6ec3b97da5680:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:809030b2249a8817c3eb0cf591ab72c8947137057f449ef2af65b6ca5a1ca455:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:5f6a107eb8ecba7e67c21fc2258afa71a9164062b2baf7e6315168a04ba275de:    done           |++++++++++++++++++++++++++++++++++++++| 
elapsed: 10.9s                                                                    total:  1.1 Gi (103.7 MiB/s)                                     
manifest-sha256:566a6a529dde04246b54d2529813c07a41aab856f90e2d4f54e976388867083d: waiting        |--------------------------------------| 
layer-sha256:c549ccf8d472c3bce9ce02e49c62b8f6cbc736ea2b8ba812a1ae9390c69d0b71:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:e206fcfcaa68aada0f94e858f8d424f91af0e0c546a729bc596bcd42b3abc3c3:    done           |++++++++++++++++++++++++++++++++++++++| 
config-sha256:94c8dc4d4b3105ad81fe1b32d3b6efcb76cdb1c504f210a418360cd607f8aa56:   done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:de26e676faa8dace0626a4b8f057d0ffc7e5139850cda2ab65b8ba0cf807a9b5:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:6c65c50510f39c06f51e7754c41825c087f5b8723aa432478e3f882dc7df5642:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:a89eb75169a1f0fb62c225c407f426d720913c741535cb49eedcb9502e2da544:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:c18636dc46c633cdc8d8950f155a7198ce1ad471661004a9d1b2f3c7cccae942:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:4f71d15c3a446a661c8e9ab34b01fe92bb2883b38da3cec9c950e6c09aa42806:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:7c5f5d979663fd14cba081a67be95348f9cc02e733f225ad804985ba2d46fb28:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:a63f5c102e3f9e02c34d75263760a275a853532c74f1d3167285f8d50f53c5a9:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:ec8b60697d022740e75072ce6c66e9f62f7c6a092adb58d67f7bcbf7073fbb4a:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:4ef8cbbd0ce3a5179cb0bf1b811c6085bfc75c884a72ab785781463ea98b853a:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:4e24d4e20fa90522001cff568936a678858181715a8ec17d95aef35d16c9ac03:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:e888fd7dc15b0eb45fd613496e238da840cdf4cfbdd87d97f450883af236a420:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:014b3810d91b80a9b7edcbce91e7645e25cd5c2c44d443f4fc6ac6552d5a7815:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:fd1d80db8223fbd21a5b7e409e67392944722a79c5c4c15f9c03545ce27fc753:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:853f1ee8487276eb37aafedb51190aa8a7d939f2d881d714a228ab9f8e1989d3:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:35a341dd8cc8619373053d7ee65e31727215c7598a7b8b7639cab5bb60ebcd6c:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:b1ab788b1035e3083f31903897e814d6fb66b3a1a06beccbd41b1d098a3c6dbb:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:0139f532dea9b33551a7c7cf03ff68992acdd50722211374ac2b4524ab3b62f2:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:81b822a9acd6ac46e4a62639e5bbb3492dcf0840bf6dbb0f144a2caf89db0af8:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:38793e44da0b76e16afca557876128b16dd3972c3db143039e585868d04ffc13:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:fa2628cfbdaabed19236bb2f5732eb05897fbe607204c3782ed6789fa804cee1:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:667d2e5ed3c8828002d782c84374d74c91a61eace367510e107f39d3d67c3b72:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:ae12f1b887c3bbba6da00f220e550bec8cca69f61b1e208066da2cfdaff6b8ba:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:4f546916dcf91371fc0f6fe56058dae34b4230c95852c0371bc6ec3b97da5680:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:809030b2249a8817c3eb0cf591ab72c8947137057f449ef2af65b6ca5a1ca455:    done           |++++++++++++++++++++++++++++++++++++++| 
layer-sha256:5f6a107eb8ecba7e67c21fc2258afa71a9164062b2baf7e6315168a04ba275de:    done           |++++++++++++++++++++++++++++++++++++++| 
elapsed: 11.0s                                                                    total:  1.1 Gi (102.7 MiB/s)                                     
ctr: failed commit on ref "manifest-sha256:566a6a529dde04246b54d2529813c07a41aab856f90e2d4f54e976388867083d": unexpected status: 404 Not Found
```

**Comment by akhilerm** [comment]
@ahgraber I am exactly not sure what the issue is here, tried searching for similar issue in containerd. Seems that since this uses an old containerd version, the bug may have been fixed in new releases. I have to try upgrading to a later version of containerd

**Comment by ahgraber** [comment]
Thanks for following up on this!

I don't have a full understanding of the nuts-and-bolts of this action (so I could be way off target here), but given that it fails on uploading the manifest and not any of the layers, is it possible the upload process only expects standard image layers and not a docker manifest?

**Comment by lwille** [comment]
I can confirm this issue when trying to retag a `gcr.io` image. The image was pushed using `docker/build-push-action@v2` in another job.
The workflow is very similar to https://github.com/docker/build-push-action/blob/master/docs/advanced/copy-between-registries.md, except that docker build and retagging are done in separate jobs.


relevant output:
```
ctr [content fetch -u file:/root/.docker/config.json --all-platforms gcr.io/*****/******:ci-34188983b0d8caea1237d7ed5580ac66a50a558c]
...
ctr [images push -u file:/root/.docker/config.json gcr.io/*****/******:*****-34188983b0d8caea1237d7ed5580ac66a50a558c gcr.io/*****/******:***** *****/******:ci-34188983b0d8caea1237d7ed5580ac66a50a558c]
...
ctr: failed commit on ref "manifest-sha256:30c0c7920b92cb0a54f8e1aa06d9e3a4a79f58f7ea0f0e06ea7d87b963b6375a": unexpected status: 404 Not Found
```

**Comment by akhilerm** [comment]
I am exploring on another solution using https://github.com/regclient/regclient, to copy images between registries. Once identified a stable one, will release a v2 for the action using the new binary.

**Comment by klutchell** [comment]
I'm seeing this issue as well when trying to copy to multiple registries, or just use multiple tags in the same registry. I've used the workaround to provide only one value for `dst` and it seems to work for now. Looking forward to a v2!

**Comment by Niek** [comment]
Same issue here when pushing two tags to the same registry - pushing a single tag now as a workaround.

**Comment by akhilerm** [comment]
Giving a heads up on what is going on this issue. 

I have updated the containerd version to 1.5.5 (latest as of now) and golang version to 1.17 and is trying out the github action. Still I am facing issues with pushing to multiple registries. If anyone wants to try out the new changes its currently updated in `akhilerm/tag-push-action@experiment` . It would be great if someone could test this version and let me know of any regression in the existing behaviour.

**Comment by JackDandy** [comment]
I also followed the example on `https://github.com/docker/build-push-action/blob/master/docs/advanced/copy-between-registries.md` to copy between docker and ghcr and same 404 error, which led me here while searching a solution.  This is the type of thing devs want to setup and forget, and given two weeks of no feedback on branch `experiment`, I suggest you push it to main so people can test it and feedback as the copy feature clearly isn't working.

**Comment by akhilerm** [comment]
The latest version of containerd which is used in the experiment has
another containerd issue which causes failures in some cases. The
containerd issue is reported here.
https://github.com/containerd/containerd/issues/5978

**Comment by akhilerm** [comment]
I have pushed a new commit 412ab72db47c0b7c69053b3bccfa4d401404efd1 into the main branch, which uses the new repo-copy image(refer [0f1843](https://github.com/akhilerm/repo-copy/commit/0f18436105522188a37f0be44624069d5c3179a3)). I have migrated from  containerd to using crane from google.

The old version which uses containerd will be now developed in the v1.x branch, where I will push the changes as soon as the bugs in containerd is fixed.

**TLDR;**
The latest version `akhilerm/tag-push-action@main` now has the fix for pushing  multiple tags at once.   It would be great, if people try it out and let me know so that the issue can be resolved/closed.

**Comment by bezysoftware** [comment]
I just tried with this result:
```
Status: Downloaded newer image for akhilerm/repo-copy:latest
crane [copy ghcr.io/bezysoftware/autosats:latest docker.io/bezysoftware/autosats:v0.0.1-test docker.io/bezysoftware/autosats:latest]
Error: parsing reference for "docker.io/bezysoftware/autosats:v0.0.1-test docker.io/bezysoftware/autosats:latest": could not parse reference: docker.io/bezysoftware/autosats:v0.0.1-test docker.io/bezysoftware/autosats:latest
panic: exit status 1

goroutine 1 [running]:
main.main()
	/go/src/github.com/akhilerm/repo-copy/main.go:20 +0x15a
Error: The process '/usr/bin/docker' failed with exit code 2
```

This is my yml:
```yml
      - name: Push image
        uses: akhilerm/tag-push-action@main
        with:
          src: ghcr.io/bezysoftware/autosats:latest
          dst: |
            docker.io/bezysoftware/autosats:${{ github.event.release.tag_name }}
            docker.io/bezysoftware/autosats:latest
```

**Comment by ahgraber** [comment]
Confirming -- the update works to push to alternate registry, but have to call once per tag

This works:

```yaml
      - name: Push image to alternate registries
        uses: akhilerm/tag-push-action@412ab72db47c0b7c69053b3bccfa4d401404efd1
        with:
          src: ${{ env.GH_REGISTRY }}/${{ env.IMAGE_NAME }}:${{ env.FULLTAG }}
          dst: |
            ${{ env.D_REGISTRY }}/${{ env.IMAGE_NAME }}:${{ env.PYTHONTAG }}

      - name: Push image to alternate registries
        uses: akhilerm/tag-push-action@412ab72db47c0b7c69053b3bccfa4d401404efd1
        with:
          src: ${{ env.GH_REGISTRY }}/${{ env.IMAGE_NAME }}:${{ env.FULLTAG }}
          dst: |
            ${{ env.D_REGISTRY }}/${{ env.IMAGE_NAME }}:${{ env.PRODTAG }}

```

This does not:

```yaml
      - name: Push image to alternate registries
        uses: akhilerm/tag-push-action@412ab72db47c0b7c69053b3bccfa4d401404efd1
        with:
          src: ${{ env.GH_REGISTRY }}/${{ env.IMAGE_NAME }}:${{ env.FULLTAG }}
          dst: |
            ${{ env.D_REGISTRY }}/${{ env.IMAGE_NAME }}:${{ env.PYTHONTAG }}
            ${{ env.D_REGISTRY }}/${{ env.IMAGE_NAME }}:${{ env.PRODTAG }}
```

**Comment by akhilerm** [comment]
@ahgraber , What is the error received , when multiple tags are being pushed at once?, Also, can you try with the `@main` version, instead of the commit. Because the commit that is referred does not contain the generated npm code.

**Comment by akhilerm** [comment]
@bezysoftware @ahgraber Sorry there was an issue in parsing multiline arguments from TS. have fixed it now.  

I have tested it in a different repository for [this](https://github.com/akhilerm/release-test/blob/main/.github/workflows/test.yml) workflow yaml, the results can be found [here](https://github.com/akhilerm/release-test/runs/3716931429?check_suite_focus=true)

**Comment by bezysoftware** [comment]
I can verify it works now, thanks!

**Comment by ahgraber** [comment]
Also confirming it works!

**Comment by akhilerm** [comment]
@ahgraber If its resolved, can you close the issue.
